### DISCUSSION TITLE
Connect receipt validation user with server notifications

### DISCUSSION THREAD

**Discussion by Iamscaredofcode** [imdhemy/laravel-in-app-purchases#24] [Q&A] [answered]
Hello @imdhemy! I am experimenting with this package and i like the way it is build - very professional!

I am wondering why not more people are using it because this seems to be the best on the market.

Altough I have one question:

Which property on the receipt should I associate with the user, so that I can fetch the user again in the server notification listeners.

In the readme i see the use of getUniqueIdentifier(), but how to I first store this value when validating the receipt?

```php
$notification = $event->getServerNotification();
$subscription = $notification->getSubscription();
$uniqueIdentifier = $subscription->getUniqueIdentifier();
$user = $this->findUserBySubscriptionId($uniqueIdentifier);
```

I hope I could describe it clearly.

Thank you

**Comment by imdhemy** [comment, marked as answer]
Hello, @Iamscaredofcode
I'm really happy to read your nice words. Thank you! ❤

I faced the same problem during the development of one of our apps. Google Play allows the developer to attach a `developer payload`, but the App Store does not.

As a workaround, I implemented an endpoint that allows the mobile application to inform the backend about the purchased subscriptions.

1. Once the subscription is purchased on the mobile application, it receives receipt data from the provider.
2. The mobile application sends the data to your backend through the previously mentioned endpoint.
3. You can find the user through your authentication process, and connect the receipt data with the authenticated user.

It's something like, hey backend, I just purchased this subscription from the IAP! 😅

**Reply by Iamscaredofcode** [reply]
Thank you for the fast answer! Actually, the way you described it is how I am doing it so far until step 3.

The question for me is how to continue after:

I have the user and I have the receipt, but what of the receipt would I have to store in order to match it with the `getUniqueIdentifier()` received from the server notification later?

I hope I could describe what I mean

**Reply by Iamscaredofcode** [reply]
Sorry, forgot to tag you @imdhemy

**Reply by imdhemy** [reply]
@Iamscaredofcode

**Tl;dr**
Use the returned value from `getUniqueIdentifier()`

---

After the purchase process is completed through and **Android Application**, a subscription token is received. This token is the same for all future subscription updates.

For the iOS apps, there will be a new transaction id for any update, but the original transaction id maintains the same.

Hence, `getUniqueIdentifier()`, is added to get a unique identifier that keeps tracks of the same subscription for different subscription states.

**Reply by Iamscaredofcode** [reply]
Sorry for the many follow ups, but I think this is missing in the documentation:

```php
// First call to verify receipt
$receipt = Subscription::appStore()->receiptData($request->input('receipt'))->verifyRenewable(); // Getting back ReceiptResponse

$user->appstore_subscription = $receipt-> // <-- WHICH VALUE TO STORE HERE FROM THE RECEIPT (there is no getUniqueIdentifier() on the receipt or receiptResponse)
$user->save();

// Server notification Event/Listener
$notification = $event->getServerNotification();
$subscription = $notification->getSubscription();
$uniqueIdentifier = $subscription->getUniqueIdentifier();

$user = User::where('appstore_subscription', $uniqueIdentifier)->first();
```

**Reply by imdhemy** [reply]
@Iamscaredofcode

You are welcome!
---

```php
$receiptResponse = Subscription::appStore()->receiptData($request->input('receipt'))->verifyRenewable();

if($receiptResponse->getStatus()->isValid()){
    $latestReceiptInfo = $receiptResponse->getLatestReceiptInfo()[0];
    $uniqueId = $latestReceiptInfo->getOriginalTransactionId();
}
```

**Reply by imdhemy** [reply]
Hi @Iamscaredofcode,
Here is a recently published update that allows you to get the unique id easier and cleaner,

```php
$verifier = Subscription::appStore()->receiptData($request->input('receipt'));
$response = $verifier->verifyRenewable();

if($response->getStatus()->isValid()){
    $uniqueId = $verifier->toStd()->getUniqueIdentifier();
}
```

The method `toStd()` is added to the Subscription facade. It works on both Google and App Store receipts to get an object which implements the [SubscriptionContract.](https://github.com/imdhemy/laravel-in-app-purchases/blob/master/src/Contracts/SubscriptionContract.php) Hope it helps.

**Reply by Iamscaredofcode** [reply]
@imdhemy I love this change! Thank you for notifying me!

**Reply by imdhemy** [reply]
@Iamscaredofcode
I decided to add that method since you created this discussion. Looking forward to your suggestions and new ideas to implement.

**Reply by Iamscaredofcode** [reply]
@Iamscaredofcode this works really good for Apple. I now try to make it work for Google ... what would be the equivalent of this for Google?

```
$verifier = Subscription::appStore()->receiptData($request->input('receipt'));
$response = $verifier->verifyRenewable();

if($response->getStatus()->isValid()){
    $uniqueId = $verifier->toStd()->getUniqueIdentifier();
}
```

**Reply by michavie** [reply]
did you make this work with google? i sometimes receive a different uniqueId/purchaseToken from the real-time notifications which prevents retrieving the user and causes the handling to fail

**Reply by imdhemy** [reply]
@michavie, yup, for both.

Did you track when the purchase token changes?

**Reply by michavie** [reply]
@imdhemy this seems to be the problem - the token is probably not invalidated correctly.

`Subscription upgrades, downgrades, and other subscription purchase flows generate purchase tokens that must replace a previous purchase token. You must invalidate the purchase tokens that appear in the linkedPurchaseToken field of the Google Play Developer API.`

Still trying to figure out the best way to invalidate old tokens. do you maybe have an example for this that we can document?

**Reply by imdhemy** [reply]
@michavie, check this topic #38

**Reply by michavie** [reply]
@imdhemy is it generally recommended to use the purchaseToken for purchase attribution, correct? (developer payload seems to become deprecated soon)

**Reply by imdhemy** [reply]
@michavie I think so!

**Reply by Iamscaredofcode** [reply]
@michavie No I didn't release my app until this day
